File: Source/Example.Chat.Server/ChatServer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

using Orleans;
using Orleankka;

namespace Example
{
    public class ChatServer : Orleankka.Typed.TypedActor
    {
        public ChatServer()
        {
            Users = new Dictionary<string, IObserverCollection>();
        }

        public Dictionary<string, IObserverCollection> Users { get; set; }

        private void NotifyClients(object message, Dictionary<string, IObserverCollection>.ValueCollection clients)
        {
            foreach (var user in clients)
            {
                user.Notify(message);
            }
        }

        public Task Join(string username, ObserverRef client)
        {
            IObserverCollection clients;
            if (Users.TryGetValue(username, out clients))
            {
                clients.Add(client);
            }
            else
            {
                IObserverCollection newClientsCollection = new ObserverCollection();
                newClientsCollection.Add(client);
                Users.Add(username, newClientsCollection);
                client.Notify(new NotificationMessage
                {
                    Text = "Hello and welcome to Orleankka chat example"
                });
            }


            NotifyClients(new NotificationMessage
            {
                Text = string.Format("User: {0} connected...", username)
            }, Users.Values);

            return TaskDone.Done;
        }

        public Task Say(string username, string text)
        {
            NotifyClients(new NewMessage
            {
                Username = username,
                Text = text
            }, Users.Values);
            return TaskDone.Done;
        }

        public Task Disconnect(string username, ObserverRef client)
        {
            IObserverCollection clients;
            if (Users.TryGetValue(username, out clients))
            {
                if (clients.Count() == 1)
                {
                    Users.Remove(username);
                }
                else
                {
                    clients.Remove(client);
                }
            }
            return TaskDone.Done;
        }
    }
}

File: Source/FSharp.Example.Observers.Chat.Server/ChatServer.fs
module ChatServer

open System.Linq
open System.Collections.Generic

open Orleankka
open Orleankka.FSharp


type ClientMessage =
   | NewMessage of Username:string * Text:string
   | Notification of Text:string

type ServerMessage =
   | Join of Username:string * Client:ObserverRef
   | Say of Username:string * Text:string   
   | Disconnect of Username:string * Client:ObserverRef

type ChatServer() = 
   inherit Actor<ServerMessage>()

   let _users = Dictionary<string, IObserverCollection>()

   let notifyClients msg = _users.Values |> Seq.iter(fun clients -> clients.Notify(msg))

   override this.Receive message = task {
      match message with
      
      | Join (userName, client) -> 
         match _users.TryGetValue(userName) with         
         | (true, clients) -> clients.Add(client)         
         | _ -> Notification(sprintf "User: %s connected..." userName) |> notifyClients
                let clients = ObserverCollection() :> IObserverCollection
                clients.Add(client)
                _users.Add(userName, clients)
                         
         return response("Hello and welcome to Orleankka chat example")
      
      | Say (userName, text) -> NewMessage(userName, text) |> notifyClients
                                return response()
      
      | Disconnect (userName, client) -> 
         match _users.TryGetValue(userName) with
         | (true, clients) -> if clients.Count() = 1 then _users.Remove(userName) |> ignore
                              else clients.Remove(client)
                              return response()
         | _ -> return response()
   }